### PR TITLE
Segment batch search

### PR DESCRIPTION
**Pull request by generall**
Another attempt to benchmark batching query.

Main question: at what level should we split batch into individual function calls?

There are 3 main options:

- On collection level - where we can simply have batch request equivalent to the list of `SearchRequest`
  - pros: Simplest to implement - no change even in internal API
  - cons: Require to perform multiple internal calls to remote shard

- On local shard level 
  - pros: 
    - No need to change code on segment level and segment holder level
    - Single request in remote shard
   - cons
     - Same filter calculated multiple times

- On segment level
  - pros:
     - Presumably smallest overhead on function calls
     - Can reuse point selection from filters
   - cons:
     - requires the most amount of changes and potential code duplication

This PR introduces benchmark to compare scenario 2 and 3.
It makes 100 single queries to 2000 stored vectors vs single batch query with 100 vectors. Dim = 100
No indexes used.
Results:

No filters:
```
batch-search-bench/search-0                                                                             
                        time:   [4.3911 ms 4.4788 ms 4.5933 ms]
batch-search-bench/search-batch-0                                                                             
                        time:   [3.7360 ms 3.8431 ms 3.9657 ms]
```

Filter which matches 1/5 of the points:
```
batch-search-bench/search-1                                                                            
                        time:   [13.916 ms 14.146 ms 14.397 ms]
batch-search-bench/search-batch-1                                                                             
                        time:   [1.1711 ms 1.1870 ms 1.2080 ms]
```

Filter which matches all data points (to estimate filter overhead):
```
batch-search-bench/search-2                                                                            
                        time:   [19.268 ms 19.852 ms 20.570 ms]
batch-search-bench/search-batch-2                                                                            
                        time:   [5.4323 ms 5.4829 ms 5.5404 ms]
```

### CODE REVIEW

**Comment by agourlay** [comment]
rebased & fixed conflicts :heavy_check_mark:

**Comment by agourlay** [comment]
## Todos

- tackle code duplication
- test batched search result vs regular search
- benchmark HNSW index search vs master
- refactor `query_points` in `struct_payload_index` to remove some Iterators

**Comment by agourlay** [comment]
We can start an official round of reviews here to get things started :hammer:

File: lib/api/src/grpc/proto/points.proto [hunk 1]
```diff
@@ -117,6 +117,11 @@ message SearchPoints {
   optional uint64 offset = 9; // Offset of the result
 }
 
+message SearchBatchPoints {
+  string collection_name = 1; // Name of the collection
```
Comment (agourlay) [inline lib/api/src/grpc/proto/points.proto:121]: The name of the collection is passed explicitly although the `search_points` contain a dedicated `collection_name` field.

The service validates that the `collection_name` exists and assumes all points in the batch targets the same collection.

This is not extremely clean for the time being.

File: lib/collection/benches/prof.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
+use std::fs::File;
```
Comment (agourlay) [inline lib/collection/benches/prof.rs:1]: Is it possible to share a single `prof.rs` file between all the `benches` across crates?

File: lib/collection/src/collection.rs [hunk 3]
```diff
@@ -642,38 +643,128 @@ impl Collection {
...
         &self,
-        request: SearchRequest,
+        request: SearchRequestBatch,
         search_runtime_handle: &Handle,
         shard_selection: Option<ShardId>,
-    ) -> CollectionResult<Vec<ScoredPoint>> {
+    ) -> CollectionResult<Vec<Vec<ScoredPoint>>> {
+        // A factor which determines if we need to use the 2-step search or not
+        // Should be adjusted based on usage statistics.
+        const PAYLOAD_TRANSFERS_FACTOR_THRESHOLD: usize = 10;
```
Comment (agourlay) [inline lib/collection/src/collection.rs:654]: the logic is adapted from the regular `search` taking in consideration the batching aspects

File: lib/collection/src/collection.rs [hunk 4]
```diff
@@ -642,38 +643,131 @@ impl Collection {
...
+                    Order::LargeBetter => {
+                        peek_top_largest_scores_iterable(res, request.limit + request.offset)
+                    }
+                    Order::SmallBetter => {
+                        peek_top_smallest_scores_iterable(res, request.limit + request.offset)
+                    }
+                };
+                // Remove offset from top result.
+                if request.offset > 0 {
+                    if top_res.len() >= request.offset {
```
Comment (agourlay) [inline lib/collection/src/collection.rs:759]: Added a fix after testing the usage of offsets.
The bug must be present in our regular search already.

**Comment by generall** [comment]
LGTM

**Review by agourlay — APPROVED** [review]
Need to approve my own work on this one.

**Pull request merged**